Python/specialize.c

```diff
@@ -778,6 +793,151 @@ _Py_Specialize_StoreAttr(PyObject *owner, _Py_CODEUNIT *instr, PyObject *name, S
+{
+    _PyAdaptiveEntry *cache0 = &cache->adaptive;
+    _PyAttrCache *cache1 = &cache[-1].attr;
+    _PyObjectCache *cache2 = &cache[-2].obj;
+
+    PyTypeObject *owner_cls = Py_TYPE(owner);
+    if (PyModule_CheckExact(owner)) {
+        int err = specialize_module_load_attr(owner, instr, name, cache0, cache1,
+            LOAD_METHOD, LOAD_METHOD_MODULE);
+        if (err) {
+            goto fail;
+        }
+        goto success;
+    }
+    if (owner_cls->tp_dict == NULL) {
+        if (PyType_Ready(owner_cls) < 0) {
+            return -1;
+        }
+    }
+    if (Py_TYPE(owner_cls)->tp_dictoffset < 0) {
+        SPECIALIZATION_FAIL(LOAD_METHOD, SPEC_FAIL_OUT_OF_RANGE);
+        goto fail;
+    }
+    PyObject **owner_dictptr = _PyObject_GetDictPtr(owner);
+    int owner_has_dict = (owner_dictptr != NULL && *owner_dictptr != NULL);
+    PyDictObject *owner_dict = owner_has_dict ? (PyDictObject *)*owner_dictptr : NULL;
+    // Check for classmethods.
+    int owner_is_class = (owner_cls->tp_getattro != PyObject_GenericGetAttr) &&
+        PyObject_TypeCheck(owner, &PyBaseObject_Type) && PyType_Check(owner);
+
+    PyObject *descr = NULL;
+    DesciptorClassification kind = 0;
+    if (owner_is_class) {
+        // Class method.
+        kind = analyze_descriptor((PyTypeObject *)owner, name, &descr, 0);
+        // Store the version right away, in case it's modified halfway through.
+        cache1->tp_version = ((PyTypeObject *)owner)->tp_version_tag;
+    }
+    else {
+        // Instance method,
+        kind = analyze_descriptor(owner_cls, name, &descr, 0);
+        cache1->tp_version = owner_cls->tp_version_tag;
+    }
+    assert(descr != NULL || kind == ABSENT || kind == GETSET_OVERRIDDEN);
+    switch (kind) {
+        case METHOD:
+            break;
+#if SPECIALIZATION_STATS
+        case NON_OVERRIDING:
+            if (Py_IS_TYPE(descr, &PyClassMethodDescr_Type)) {
+                // Builtin METH_CLASS class method -- ie wrapped PyCFunction.
+                // (KJ): Don't bother, rare and no speedup in microbenchmarks.
+                SPECIALIZATION_FAIL(LOAD_METHOD, SPEC_FAIL_BUILTIN_CLASS_METHOD);
+            }
+            else if (Py_IS_TYPE(descr, &PyClassMethod_Type)) {
+                // This is the actual Python classmethod(func) object.
+                SPECIALIZATION_FAIL(LOAD_METHOD, SPEC_FAIL_CLASS_METHOD_OBJ);
+            }
+        goto fail;
+#endif
+    default:
+        SPECIALIZATION_FAIL(LOAD_METHOD, SPEC_FAIL_NOT_METHOD);
+        goto fail;
+    }
+    
+    assert(kind == METHOD);
+    // Technically this is fine for bound method calls, but slightly slower at
+    // runtime to get dict. TODO: profile pyperformance and see if it's worth it
+    // to slightly slow down the common case, so that we can specialize this
+    // uncommon one.
+    if (owner_cls->tp_dictoffset < 0) {
+     SPECIALIZATION_FAIL(LOAD_METHOD, SPEC_FAIL_NEGATIVE_DICTOFFSET);
+     goto fail;
```

Comment:
@markshannon we could push `LOAD_METHOD` cache hits from 92.5% to 95% if we used the generic ``_PyObject_GetDictPtr`` to support negative offsets.

However, that would slightly slow down the other ``90%`` where ``dictoffset >=0``. So I have zero clue if it's worth it.

Reply:
It isn't worth it. We should fix object layout instead https://github.com/faster-cpython/ideas/issues/80

Reply:
Cool, got it. I'm excited; that looks like it'll play very nicely with all the LOAD/STORE specializations so far.